drivers/platform/mbed/mbed_gpio.cpp

```diff
@@ -0,0 +1,308 @@
+/******************************************************************************/
+
+/**
+ * @brief Obtain the GPIO descriptor.
+ * @param desc[in,out] - The GPIO descriptor.
+ * @param param[in] - The structure that contains the GPIO init parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_gpio_get(struct gpio_desc **desc,
+		      const struct gpio_init_param *param)
+{
+	struct gpio_desc *gpio_desc;
+	struct mbed_gpio_desc *mbed_gpio_desc;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	/* Create gpio description object for the device */
+	gpio_desc = (struct gpio_desc *)calloc(1, sizeof(*gpio_desc));
+	if (!gpio_desc)
+		return -ENOMEM;
+
+	gpio_desc->number = param->number;
+
+	/* Create the gpio extra descriptor object to store extra mbed gpio info */
+	mbed_gpio_desc = (struct mbed_gpio_desc *)calloc(1,
+			 sizeof(*mbed_gpio_desc));
+	if (!mbed_gpio_desc)
+		goto err_mbed_gpio_desc;
+
+	mbed_gpio_desc->direction = GPIO_IN;
+	mbed_gpio_desc->gpio_pin = NULL;
+
+	if (param->extra)
+		mbed_gpio_desc->pin_mode = ((struct mbed_gpio_init_param *)
+					    param->extra)->pin_mode;
+	else
+		mbed_gpio_desc->pin_mode = NULL;
+
+	gpio_desc->extra = mbed_gpio_desc;
+	*desc = gpio_desc;
+
+	return SUCCESS;
+
+err_mbed_gpio_desc:
+	free(gpio_desc);
+
+	return -ENOMEM;
+}
+
+/**
+ * @brief Get the value of an optional GPIO.
+ * @param desc[in,out] - The GPIO descriptor.
+ * @param param[in] - GPIO Initialization parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_gpio_get_optional(struct gpio_desc **desc,
+			       const struct gpio_init_param *param)
+{
+	if (param == NULL) {
+		*desc = NULL;
+		return SUCCESS;
+	}
+
+	return mbed_gpio_get(desc, param);
+}
+
+/**
+ * @brief Free the resources allocated by mbed_gpio_get().
+ * @param desc[in] - The GPIO descriptor.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_gpio_remove(struct gpio_desc *desc)
+{
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	/* Free the gpio object */
+	if (((struct mbed_gpio_desc *)(desc->extra))->gpio_pin)
+		free(((struct mbed_gpio_desc *)(desc->extra))->gpio_pin);
+
+	/* Free the gpio descriptor objects */
+	free(desc->extra);
+	free(desc);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Enable the input direction of the specified GPIO.
+ * @param desc[in] - The GPIO descriptor.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ * @note does not support reconfiguration of already set pin direction
+ */
+int32_t mbed_gpio_direction_input(struct gpio_desc *desc)
+{
+	DigitalIn *gpio_input;				// pointer to gpio input object
+	struct mbed_gpio_desc
+		*gpio_desc_extra; 	// pointer to gpio desc extra parameters
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	gpio_desc_extra = (struct mbed_gpio_desc *)(desc->extra);
+	if (gpio_desc_extra->gpio_pin == NULL) {
```

Comment:
Suggestion: I think we could improve the code if we check `if (gpio_desc_extra->gpio_pin != NULL)` and do a `delete gpio_desc_extra->gpio_pin;` if we have instantiated it before. This can make the code easier to read, in my opinion, and can improve functionality by allowing us to change the direction of the pin after we set it once.

This change is not necessary if you think it doesn't make sense. Let me know.

Reply:
May also apply to `mbed_gpio_direction_output`.

Reply:
Agreed. I will update code to simplify it.